apps/nsq_to_file/file_logger.go

```diff
@@ -227,21 +266,33 @@ func (f *FileLogger) updateFile() {
 	f.lastFilename = filename
 	f.lastOpenTime = time.Now()
 
-	fullPath := path.Join(*outputDir, filename)
-	dir, _ := filepath.Split(fullPath)
-	if dir != "" {
-		err := os.MkdirAll(dir, 0770)
-		if err != nil {
-			log.Fatalf("ERROR: %s Unable to create %s", err, dir)
-		}
-	}
+	fullPath := path.Join(*workDir, filename)
+	makeOutputDir(fullPath)
 
 	f.Close()
 
 	var err error
 	var fi os.FileInfo
 	for ; ; f.rev++ {
 		absFilename := strings.Replace(fullPath, "<REV>", fmt.Sprintf("-%06d", f.rev), -1)
+
+		// If we're using a working directory for in-progress files,
+		// proactively check for duplicate file names in the output dir to
+		// prevent conflicts on rename in the normal case
+		if *workDir != *outputDir {
+			outputFileName := makeOutputPath(absFilename, *workDir, *outputDir)
+			makeOutputDir(outputFileName)
+
+			_, err := os.Stat(outputFileName)
+			if err == nil || os.IsExist(err) {
```

Comment:
I would guess that you should never get an error here (including an Exist error). But this seems very safe, and clear.

Reply:
Yeah, during normal operation this conditional should never be true.

The `os.IsExist` case covers an unlikely but possible scenario (I think!) where some entirely other process adds a file to the final output dir that has the same name as an in-progress file in the working dir, in a sequence of events along these lines:
 - Two (or more) nsq_to_file processes are sharing the same working and output dirs
 - One process is `kill -9`'d, orphaning a file in the working dir
 - A separate cleanup process scans the working dir for orphaned files and moves the file into a tmp dir for cleanup/processing
 - The killed nsq_to_file process starts back up and picks the same file name as the orphaned file (because it's no longer present in either work dir or output dir)
 - The cleanup process finishes processing and naively (or _buggily_) sticks the orphaned file into the output dir under the same name
- The restarted nsq_to_file process now has to safely pick a new filename at the point where it's trying to move the file from working dir to output dir

I think the sequence above is quite unlikely, but, as you said, I think guarding against it isn't too onerous!

Reply:
I just mean, if the file exists, I suspect the stat will never return an error, it will be the `err == nil` case.